Sources/NIOHPACK/SimpleRingBuffer.swift

```diff
@@ -0,0 +1,609 @@
+                assert(idx == underestimatedCount)
+                remainderWritten = try _setRemainder(iterator: iterator, at: index + idx, available: self.writableBytes - idx)
+            } else {
+                // first part...
+                let firstPart = ptr.count - index
+                let secondPart = underestimatedCount - firstPart
+                
+                // split into two sub-sequences that won't go bang inside _copyMemory() if the available target memory is
+                // less than the sequence's 'underestimatedCount'.
+                let firstSubSequence = bytes.dropLast(secondPart)
+                let secondSubSequence = bytes.dropFirst(firstPart)
+                
+                let (_, firstIdx) = UnsafeMutableBufferPointer(start: base, count: firstPart).initialize(from: firstSubSequence)
+                // second part...
+                base = UnsafeMutablePointer(mutating: ptr.baseAddress!.assumingMemoryBound(to: UInt8.self))
+                let (iterator, idx) = UnsafeMutableBufferPointer(start: base, count: secondPart).initialize(from: secondSubSequence)
+                assert(firstIdx + idx == underestimatedCount)
+                remainderWritten = try _setRemainder(iterator: iterator, at: idx, available: ptr.count - firstIdx + idx)
+            }
+            
+            return underestimatedCount + remainderWritten
+        }
+    }
+    
+    // MARK: Public Core API
+    
+    init(allocator: ByteBufferAllocator, capacity: Int) {
+        self._storage = allocator.buffer(capacity: capacity)
+    }
+    
+    /// The number of bytes available to read.
+    public var readableBytes: Int {
+        // we maintain readable (used) byte count separately because readIdx == writeIdx could mean either empty or full.
+        return Int(self._readableBytes)
+    }
+    
+    /// The number of available bytes for writing.
+    public var writableBytes: Int {
+        return self.capacity - self.readableBytes
+    }
+    
+    /// The total capacity of this `CircularByteBuffer`.
+    public var capacity: Int {
+        return Int(self._storage.capacity)
+    }
+    
+    /// Change the capacity to exactly `to` bytes.
+    ///
+    /// - parameters:
+    ///     - to: The desired minimum capacity.
+    /// - precondition: Int cannot shrink beyond the number of bytes
+    ///                 currently in the buffer.
+    public mutating func changeCapacity(to newCapacity: Int) {
+        precondition(newCapacity >= self.readableBytes,
+                     "new capacity \(newCapacity) is too small to contain current buffer contents")
+        if newCapacity == self._storage.capacity {
+            return
+        }
+        
+        // we know this capacity is ok
+        self._reallocateStorageAndRebase(capacity: newCapacity)
+    }
+    
+    public mutating func makeContiguous() {
```

Comment:
I think this function is used only in testing. Do we need it at all for those tests? If we do, let's make it internal, if we don't, let's remove it entirely.